File: paleo_workbench/ui/map_action_controller.py
"""Central QAction state for the GIS authoring workspace."""

from __future__ import annotations

from PySide6.QtCore import QObject, QSize, Qt, Signal
from PySide6.QtGui import QAction, QActionGroup, QIcon, QKeySequence
from PySide6.QtWidgets import QToolBar, QWidget

from paleo_workbench.mapping.action_registry import ACTION_SPECS
from paleo_workbench.mapping.tool_help import (
    TOOL_LABELS,
    TOOL_SHORTCUTS,
)

__all__ = ["MapActionController"]


def _map_icon(action_id: str, *, fallback: str = "") -> QIcon:
    """Load a toolbar icon (map/ 优先，assets 根目录兜底)，缺失返回空 QIcon。

    E2：经主题感知 tint 工厂加载——中性灰色字形按当前主题 TEXT_SECONDARY
    重染；多色/语义色 SVG 原样返回（拉平会破坏图义）。tint 工厂懒加载：
    workstation 包 init 经 composite_document 模块级导入本模块，顶层
    import common 会成环。
    """
    from paleo_workbench.ui.workstation.common import tinted_map_icon

    for name in (action_id, fallback):
        if not name:
            continue
        icon = tinted_map_icon(name)
        if not icon.isNull():
            return icon
    return QIcon()


class MapActionController(QObject):
    """One source of QAction checked/enabled state across menus and toolbars."""

    tool_requested = Signal(str)
    command_requested = Signal(str)

    _TOOL_IDS = (
        "pan", "zoom_in", "zoom_out", "identify", "select", "select_rectangle",
        "measure_distance", "add_point", "add_line", "add_polygon", "move_feature", "vertex",
        "reshape", "add_ring", "add_part",
    )

    #: 命令面动作（区别于画布 MapTool：触发一次命令，不置 current_tool）。
    #: V10（#1256）：提为类常量——图标资产完整性与"无 surface-only 幽灵
    #: 动作"两条断言都要枚举这份清单，内联在 _build_actions 里无法审计。
    _COMMAND_IDS = (
        "full_extent", "previous_extent", "next_extent", "refresh",
        "clear_selection", "select_all", "invert_selection", "toggle_editing",
        "save_edits", "rollback", "delete_selected",
        "undo", "redo", "split", "merge",
        "duplicate_selected", "explode_multipart", "collect_multipart",
        "snapping", "topology", "cancel",
    )

    #: 词表单一来源（V8 M4：action_help.TOOL_LABELS；帮助/QAction 同名）。
    _LABELS = dict(TOOL_LABELS)

    #: 工具条第三段动作（layer/symbology/factor/qa/layout 组的扩展面命令
    #: + geometry 组的 repair_geometry）。V10 Milestone B：图标/身份登记
    #: 单一来源 = mapping.action_registry；此处只保留**构建顺序**。
    _SURFACE_EXTENSION_IDS = (
        "layer_new", "reference_import", "layer_properties", "attribute_table",
        "layer_zoom", "layer_export", "symbology", "style_manager",
        "factor_workbench", "factor_overlay", "qa_run", "map_product_assemble",
        "map_export", "repair_geometry",
    )

    #: id → svg 名（登记处派生；缺省回落 id 本名）。
    _SURFACE_ICONS = {
        tool_id: spec.icon for tool_id, spec in ACTION_SPECS.items()
    }

    def __init__(self, parent: QObject | None = None):
        super().__init__(parent)
        self.actions: dict[str, QAction] = {}
        self._tool_group = QActionGroup(self)
        self._tool_group.setExclusive(True)
        self._build_actions()

    def _action(self, action_id: str, *, checkable: bool = False, shortcut: str = "") -> QAction:
        icon_name = self._SURFACE_ICONS.get(action_id, action_id)
        action = QAction(_map_icon(icon_name), self._LABELS[action_id], self)
        action.setObjectName(f"MapAction:{action_id}")
        action.setCheckable(checkable)
        action.setToolTip(self._LABELS[action_id])
        action.setStatusTip(self._LABELS[action_id])
        if shortcut:
            action.setShortcut(QKeySequence(shortcut))
            # The window also binds Ctrl+S to project save. Keep this mapping
            # action's shortcut confined to the editing widget tree so the two
            # don't collide into an ambiguous "Ctrl+S" shortcut.
            action.setShortcutContext(Qt.ShortcutContext.WidgetWithChildrenShortcut)
        self.actions[action_id] = action
        return action

    def _build_actions(self) -> None:
        for action_id in self._TOOL_IDS:
            action = self._action(action_id, checkable=True)
            self._tool_group.addAction(action)
            action.triggered.connect(lambda checked=False, name=action_id: checked and self.tool_requested.emit(name))
        # 快捷键单一来源（V8 M4：action_help.TOOL_SHORTCUTS；帮助镜像同源）。
        shortcut_registry = dict(TOOL_SHORTCUTS)
        for action_id in self._COMMAND_IDS:
            shortcut = shortcut_registry.get(action_id, "")
            action = self._action(action_id, checkable=action_id in {"snapping", "topology", "toggle_editing"}, shortcut=shortcut)
            action.triggered.connect(lambda checked=False, name=action_id: self.command_requested.emit(name))
        self.actions["pan"].setChecked(True)
        # V7 专业分组扩展（Layer / Symbology / Factor / QA / Layout·Export）。
        for action_id in self._SURFACE_EXTENSION_IDS:
            action = self._action(action_id)
            action.triggered.connect(lambda checked=False, name=action_id: self.command_requested.emit(name))

    def apply_availability(self, availability, *, help_texts=None) -> None:
        """Apply ``{tool_id: ToolAvailability}`` from the canonical evaluator.

        V8 M1：输入是 ``mapping.tool_availability.ToolAvailability``（唯一
        契约）——enabled/visible/checked 与禁用原因全部来自单一求值器，
        本方法只做 Qt 呈现（图标/tooltip/statusTip/勾选），不做业务判断。

        V8 M4：``help_texts`` 为 ``{tool_id: (tooltip, statusTip)}`` 覆盖
        （宿主用 ``action_help`` 从同一契约派生的解释文本）；缺省回落
        名称+原因两行式。
        """
        for tool_id, result in dict(availability).items():
            action = self.actions.get(tool_id)
            if action is None:
                continue  # evaluator may cover tools this host has no action for
            label = self._LABELS.get(tool_id, tool_id)
            action.setEnabled(bool(result.enabled))
            action.setVisible(bool(result.visible))
            override = (help_texts or {}).get(tool_id)
            if override is not None:
                tooltip, status_tip = override
                action.setToolTip(tooltip)
                action.setStatusTip(status_tip)
            else:
                reason = result.disabled_reason
                if reason:
                    action.setToolTip(f"{label}\n{reason}")
                    action.setStatusTip(f"{label}（{reason}）")
                else:
                    action.setToolTip(label)
                    action.setStatusTip(label)
            if action.isCheckable() and action.isChecked() != bool(result.checked):
                action.blockSignals(True)
                action.setChecked(bool(result.checked))
                action.blockSignals(False)

    def toolbar(
        self,
        title: str,
        action_ids: tuple[str | tuple[str, ...], ...],
        parent: QWidget | None = None,
    ) -> QToolBar:
        """Build an icon-only toolbar.

        ``action_ids`` entries are either a single action id or a nested tuple
        of ids forming a logical group; a separator is inserted between groups.
        Flat tuples of ids remain supported for compatibility.
        """
        toolbar = QToolBar(title, parent)
        toolbar.setObjectName(f"MapToolbar:{title.replace(' ', '')}")
        toolbar.setMovable(False)
        toolbar.setToolButtonStyle(Qt.ToolButtonStyle.ToolButtonIconOnly)
        toolbar.setIconSize(QSize(18, 18))
        for index, entry in enumerate(action_ids):
            ids = entry if isinstance(entry, tuple) else (entry,)
            if index and isinstance(entry, tuple):
                toolbar.addSeparator()
            for action_id in ids:
                toolbar.addAction(self.actions[action_id])
        return toolbar

File: libs/ui_controllers/src/qt/map_action_controller.cpp
#include <pwb/ui_controllers/qt/map_action_controller.hpp>

#include <QKeySequence>
#include <QSignalBlocker>
#include <QSize>

#include <algorithm>
#include <filesystem>

namespace pwb::ui_controllers::qt {

namespace {

namespace fs = std::filesystem;

fs::path g_icon_root;  // set_icon_root; empty → icon lookup misses honestly

// _resolve_icon parity: map/ first, root fallback; .svg suffix implied.
fs::path resolve_icon_path(const std::string& stem) {
    if (g_icon_root.empty()) return {};
    for (const fs::path& dir : {g_icon_root / "map", g_icon_root}) {
        const fs::path candidate = dir / (stem + ".svg");
        std::error_code ec;
        if (fs::exists(candidate, ec) && !ec) return candidate;
    }
    return {};
}

const MapActionSpec* spec_of(const std::string& action_id) {
    const auto it = action_specs().find(action_id);
    return it != action_specs().end() ? &it->second : nullptr;
}

bool is_checkable_command(const std::string& action_id) {
    const auto& ids = map_checkable_command_ids();
    return std::find(ids.begin(), ids.end(), action_id) != ids.end();
}

}  // namespace

// The icon root is host-configured (Python resolves the package assets
// dir statically); the integration adapter points it at the deployed
// assets/icons directory. Unset → icons resolve to null, same as a
// missing asset.
void set_map_icon_root(const std::filesystem::path& root) {
    g_icon_root = root;
}

QIcon map_action_icon(const std::string& icon_name,
                      const std::string& fallback) {
    for (const std::string* name : {&icon_name, &fallback}) {
        if (name->empty()) continue;
        const fs::path path = resolve_icon_path(*name);
        if (!path.empty()) {
            return QIcon(QString::fromStdString(path.generic_string()));
        }
    }
    return QIcon();
}

MapActionController::MapActionController(QObject* parent)
    : QObject(parent) {
    tool_group_ = new QActionGroup(this);
    tool_group_->setExclusive(true);

    // _TOOL_IDS — checkable canvas MapTools in one exclusive group;
    // tool_requested fires only when the trigger lands checked.
    for (const auto& tool_id : map_tool_ids()) {
        QAction* action = make_action_(tool_id, /*checkable=*/true, {});
        tool_group_->addAction(action);
        connect(action, &QAction::triggered, this,
                [this, tool_id](bool checked) {
                    if (checked)
                        emit tool_requested(QString::fromStdString(tool_id));
                });
    }
    // _COMMAND_IDS — one-shot commands (checkable subset keeps state).
    for (const auto& command_id : map_command_ids()) {
        const MapActionSpec* spec = spec_of(command_id);
        QAction* action = make_action_(
            command_id, is_checkable_command(command_id),
            spec != nullptr ? spec->shortcut : std::string{});
        connect(action, &QAction::triggered, this,
                [this, command_id](bool) {
                    emit command_requested(
                        QString::fromStdString(command_id));
                });
    }
    if (auto* pan = action("pan")) pan->setChecked(true);
    // _SURFACE_EXTENSION_IDS — V7 extension commands appended after the
    // core set (build order is the contract).
    for (const auto& action_id : map_surface_extension_ids()) {
        QAction* action = make_action_(action_id, /*checkable=*/false, {});
        connect(action, &QAction::triggered, this,
                [this, action_id](bool) {
                    emit command_requested(
                        QString::fromStdString(action_id));
                });
    }
}

QAction* MapActionController::action(const std::string& action_id) const {
    const auto it = actions_.find(action_id);
    return it != actions_.end() ? it->second : nullptr;
}

QAction* MapActionController::make_action_(
    const std::string& action_id, bool checkable,
    const std::string& shortcut) {
    const MapActionSpec* spec = spec_of(action_id);
    // Icon chain parity: spec.icon (covers the extension ids' surface
    // icons) → action_id fallback → empty icon on miss.
    const std::string icon_name =
        (spec != nullptr && !spec->icon.empty()) ? spec->icon : action_id;
    const std::string label =
        (spec != nullptr && !spec->label.empty()) ? spec->label : action_id;
    auto* action = new QAction(
        map_action_icon(icon_name), QString::fromStdString(label), this);
    action->setObjectName(
        QStringLiteral("MapAction:%1").arg(QString::fromStdString(action_id)));
    action->setCheckable(checkable);
    action->setToolTip(QString::fromStdString(label));
    action->setStatusTip(QString::fromStdString(label));
    if (!shortcut.empty()) {
        action->setShortcut(QKeySequence(QString::fromStdString(shortcut)));
        // The window also binds Ctrl+S to project save: keep this action's
        // shortcut confined to the editing widget tree (WidgetWithChildren
        // parity) so the two don't collide.
        action->setShortcutContext(
            Qt::ShortcutContext::WidgetWithChildrenShortcut);
    }
    actions_[action_id] = action;
    return action;
}

void MapActionController::apply_availability(
    const std::map<std::string, tool_policy::ToolAvailability>& availability,
    const std::map<std::string, std::pair<std::string, std::string>>*
        help_texts) {
    for (const auto& [tool_id, result] : availability) {
        QAction* target = action(tool_id);
        if (target == nullptr) {
            continue;  // evaluator may cover tools this host has no action for
        }
        std::optional<std::pair<std::string, std::string>> override_text;
        if (help_texts != nullptr) {
            const auto it = help_texts->find(tool_id);
            if (it != help_texts->end()) override_text = it->second;
        }
        const MapActionPresentation plan =
            availability_text(tool_id, result, override_text);
        target->setEnabled(plan.enabled);
        target->setVisible(plan.visible);
        target->setToolTip(QString::fromStdString(plan.tooltip));
        target->setStatusTip(QString::fromStdString(plan.status_tip));
        if (target->isCheckable() &&
            target->isChecked() != plan.checked) {
            // Checkable updates never re-emit (blockSignals parity).
            const QSignalBlocker blocker(target);
            target->setChecked(plan.checked);
        }
    }
}

QToolBar* MapActionController::toolbar(
    const QString& title,
    const std::vector<std::vector<std::string>>& action_id_entries,
    QWidget* parent) {
    auto* bar = new QToolBar(title, parent);
    bar->setObjectName(QStringLiteral("MapToolbar:%1")
                           .arg(QString(title).remove(' ')));
    bar->setMovable(false);
    bar->setToolButtonStyle(Qt::ToolButtonStyle::ToolButtonIconOnly);
    bar->setIconSize(QSize(18, 18));
    const auto plan = toolbar_plan(action_id_entries);
    for (const auto& entry : plan) {
        if (entry.separator) {
            bar->addSeparator();
            continue;
        }
        QAction* target = action(entry.action_id);
        if (target != nullptr) bar->addAction(target);
    }
    return bar;
}

}  // namespace pwb::ui_controllers::qt
